backend/src/main/java/pt/ulisboa/tecnico/socialsoftware/tutor/statement/StatementService.java

```diff
@@ -110,6 +112,54 @@ public StatementQuizDto generateStudentQuiz(int userId, int executionId, Stateme
         return new StatementQuizDto(quizAnswer);
     }
 
+    @Retryable(
+            value = { SQLException.class },
+            backoff = @Backoff(delay = 2000))
+    @Transactional(isolation = Isolation.READ_COMMITTED)
+    public StatementQuizDto generateTournamentQuiz(int userId, int executionId, StatementTournamentCreationDto quizDetails, Tournament tournament) {
+        User user = userRepository.findById(userId).orElseThrow(() -> new TutorException(USER_NOT_FOUND, userId));
+
+        Quiz quiz = new Quiz();
+        quiz.setKey(quizService.getMaxQuizKey() + 1);
+        quiz.setType(Quiz.QuizType.GENERATED.toString());
+        quiz.setCreationDate(DateHandler.now());
+
+        CourseExecution courseExecution = courseExecutionRepository.findById(executionId).orElseThrow(() -> new TutorException(COURSE_EXECUTION_NOT_FOUND, executionId));
+
+        List<Question> availableQuestions = questionRepository.findAvailableQuestions(courseExecution.getCourse().getId());
+
+        if (quizDetails.getTopicConjunction() != null) {
+            availableQuestions = filterByTopicsTournament(availableQuestions, quizDetails, courseExecution);
```

Comment:
can it be converted to a stream that filters each available question by the quizDetailTopics and the courseExecutionAvailable topics?